apps/experiments/models.py

```diff
@@ -512,6 +512,36 @@ def get_experiments_for_display(self):
             .distinct()
         )
 
+    @transaction.atomic()
+    def update_memory(self, data: dict, experiment: Experiment | None = None):
+        """
+        Updates this participant's data records by merging `data` with the existing data. By default, data for all
+        experiments the this participant participated in will be updated. If there are no records for a specific
+        experiment, one will be created.
+
+        Paramters
+        data:
+            A dictionary containing the new data
+        experiment:
+            If specified, only the data for this experiment will be updated
+        """
+        experiments = Experiment.objects.filter(team=self.team).prefetch_related("participant_data").all()
+        if experiment:
+            experiments = experiments.filter(id=experiment.id)
+
+        records_to_update = []
+        for experiment in experiments:
+            participant_data = experiment.participant_data.filter(participant=self).first()
```

Comment:
Doing a filter on `participant_data` runs a new query and won't use the prefetch cache. What you can do is filter the prefetch to only fetch data for this participant:

```python
.prefetch_related(Prefetch("participant_data", queryset=ParticipantData.objects.filter(participant=self)))
```

That should result in all objects in `experiment.participant_data` being for the current participant

Reply:
beautiful, thanks Simon!